include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+
+/* Keep these alphabetized wrt property name */
+
+/** Runtime Dynamic Battery Parameters */
+
+/** Represents the health of the charger */
+#define CHARGER_HEALTH          0
+/** Indicates if VBUS is present for the charger */
+#define CHARGER_ONLINE          CHARGER_HEALTH + 1
+/** Reports whether or not a battery is present */
+#define CHARGER_PRESENT         CHARGER_ONLINE + 1
+/** Represents the charging status of the charger */
+#define CHARGER_STATUS          CHARGER_PRESENT + 1
+
+/** Reserved to demark end of common charger properties */
+#define CHARGER_COMMON_COUNT    CHARGER_STATUS + 1
+/**
+ * Reserved to demark downstream custom properties - use this value as the actual value may
+ * change over future versions of this API
+ */
+#define CHARGER_CUSTOM_BEGIN    CHARGER_COMMON_COUNT + 1
+
+/** Reserved to demark end of valid enum properties */
+#define CHARGER_PROP_MAX        UINT16_MAX
+
+enum charger_health {
+	CHARGER_HEALTH_UNKNOWN = 0,
+	CHARGER_HEALTH_COLD,
+	CHARGER_HEALTH_OVERHEAT,
+	CHARGER_HEALTH_GOOD,
+};
+
+enum charger_online {
+	CHARGER_ONLINE_OFFLINE = 0,
+	CHARGER_ONLINE_FIXED,
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+enum charger_status {
+	CHARGER_STATUS_UNKNOWN = 0,
+	CHARGER_STATUS_CHARGING,
+	CHARGER_STATUS_DISCHARGING,
+	CHARGER_STATUS_NOT_CHARGING,
+	CHARGER_STATUS_FULL,
+};
+
+struct charger_get_property {
+	/** Battery charger property to get */
+	uint16_t property_type;
+
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int err;
+
+	/** Property field for getting */
+	union {
+		/* Fields have the format: */
+		/* CHARGER_PROPERTY_FIELD */
+		/* type property_field; */
+
+		/* Dynamic Charger Info */
+	/** CHARGER_HEALTH */
+	enum charger_health health;
+	/** CHARGER_ONLINE */
+	enum charger_online online;
+	/** CHARGER_PRESENT */
+	bool present;
+	/** CHARGER_STATUS */
+	enum charger_status status;
```

Comment:
A lot of the fuel gauge commands are supposed to have `SBS` in them when they're absolutely specific to SBS. That's a change I should have up this week with a tiny driver migration.

I think it makes sense for the charger API to follow that pattern as well where if the property is especially specific to SBS let's just put SBS in the property.

Reply:
This wasn't intended to be SBS specific, but rather akin the POWER_SUPPLY_PROP_STATUS.

I feel it is generic for a charger API to have charging, discharging, not charging, and full as the possible states.